File: C#/WindowsPhone/LGame-XNA-lib/Loon.Foundation/NSData.cs
using Loon.Net;
using Loon.Utils;
using Loon.Core;
using System.IO;
using System.Text;
namespace Loon.Foundation {
	
	public class NSData : NSObject {
	
		private byte[] bytes;
	
		public NSData(byte[] b) {
			if (Base64Coder.IsArrayByteBase64(b)) {
				bytes = Base64Coder.Decode(b);
			} else {
				this.bytes = b;
			}
		}
	
		public NSData(string base64) {
			string data = "";
			if (Base64Coder.IsBase64(base64)) {
				foreach (string line  in  Loon.Utils.StringUtils.Split(base64, "\n")) {
					data += line.Trim();
				}
				char[] enc = data.ToCharArray();
				bytes = Base64Coder.DecodeBase64(enc);
			} else {
                this.bytes = Loon.Utils.StringUtils.GetBytes(base64);
			}
		}
	
		public byte[] Bytes() {
			return bytes;
		}
	
		public int Length() {
			return bytes.Length;
		}

        public void GetBytes(Loon.Java.ByteBuffer buf, int length)
        {
            sbyte[] buffer = new sbyte[bytes.Length];
            for (int i = 0; i < buffer.Length; i++)
            {
                buffer[i] = (sbyte)bytes[i];
            }
            buf.Put(buffer, 0, MathUtils.Min(bytes.Length, length));
        }

        public void GetBytes(Loon.Java.ByteBuffer buf, int rangeStart, int rangeEnd)
        {
            sbyte[] buffer = new sbyte[bytes.Length];
            for (int i = 0; i < buffer.Length; i++)
            {
                buffer[i] = (sbyte)bytes[i];
            }
            buf.Put(buffer, rangeStart, MathUtils.Min(bytes.Length, rangeEnd));
        }
	
		public string GetBase64() {
			byte[] buffer = Base64Coder.Encode(bytes);
			try {
				return StringUtils.GetString(buffer);
			} catch (IOException) {
                return StringUtils.GetString(buffer);
			}
		}
	
		public string GetString() {
            return GetString(System.Text.Encoding.UTF8);
		}
	
		public string GetString(System.Text.Encoding format) {
			byte[] buffer = this.bytes;
			try {
				return StringUtils.GetString(buffer,format);
			} catch (IOException) {
                return StringUtils.GetString(buffer);
			}
		}

        public override bool Equals(object obj)
        {
            return obj.GetType().Equals(GetType())
                    && CollectionUtils.Equals(((NSData)obj).bytes, bytes);
        }
	
		public override int GetHashCode() {
			int hash = 5;
			hash = 67 * hash + this.bytes.GetHashCode();
			return hash;
		}
	
		protected internal override void AddSequence(StringBuilder sbr, string indent) {
			sbr.Append(indent);
			sbr.Append("<data>");
			sbr.Append(LSystem.LS);
			string base64 = GetBase64();
			foreach (string line  in  StringUtils.Split(base64, LSystem.LS)) {
				sbr.Append(indent);
				sbr.Append("  ");
				sbr.Append(line);
				sbr.Append(LSystem.LS);
			}
			sbr.Append(indent);
			sbr.Append("</data>");
		}
	
	}
}

File: Java/Android/OpenGLES/source/core/src/loon/foundation/NSData.java
/**
 * Copyright 2013 The Loon Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package loon.foundation;

import java.nio.ByteBuffer;
import java.util.Arrays;

import loon.core.LSystem;
import loon.net.Base64Coder;
import loon.utils.MathUtils;

public class NSData extends NSObject {

	private byte[] bytes;

	public NSData(byte[] b) {
		if (Base64Coder.isArrayByteBase64(b)) {
			bytes = Base64Coder.decode(b);
		} else {
			this.bytes = b;
		}
	}

	public NSData(String base64) {
		String data = "";
		if (Base64Coder.isBase64(base64)) {
			for (String line : base64.split("\n")) {
				data += line.trim();
			}
			char[] enc = data.toCharArray();
			bytes = Base64Coder.decodeBase64(enc);
		} else {
			this.bytes = base64.getBytes();
		}
	}

	public byte[] bytes() {
		return bytes;
	}

	public int length() {
		return bytes.length;
	}

	public void getBytes(ByteBuffer buf, int length) {
		buf.put(bytes, 0, MathUtils.min(bytes.length, length));
	}

	public void getBytes(ByteBuffer buf, int rangeStart, int rangeEnd) {
		buf.put(bytes, rangeStart, MathUtils.min(bytes.length, rangeEnd));
	}

	public String getBase64() {
		byte[] buffer = Base64Coder.encode(bytes);
		try {
			return new String(buffer, LSystem.encoding);
		} catch (java.io.UnsupportedEncodingException uue) {
			return new String(buffer);
		}
	}

	public String getString() {
		return getString(LSystem.encoding);
	}

	public String getString(String format) {
		byte[] buffer = this.bytes;
		try {
			return new String(buffer, format);
		} catch (java.io.UnsupportedEncodingException uue) {
			return new String(buffer);
		}
	}

	@Override
	public boolean equals(Object obj) {
		return obj.getClass().equals(getClass())
				&& Arrays.equals(((NSData) obj).bytes, bytes);
	}

	@Override
	public int hashCode() {
		int hash = 5;
		hash = 67 * hash + Arrays.hashCode(this.bytes);
		return hash;
	}

	@Override
	protected void addSequence(StringBuilder sbr,String indent) {
		sbr.append(indent);
		sbr.append("<data>");
		sbr.append(LSystem.LS);
		String base64 = getBase64();
		for (String line : base64.split(LSystem.LS)) {
			sbr.append(indent);
			sbr.append("  ");
			sbr.append(line);
			sbr.append(LSystem.LS);
		}
		sbr.append(indent);
		sbr.append("</data>");
	}

}
